sound/soc/sof/sof-audio.c

```diff
@@ -408,25 +444,6 @@ int sof_widget_list_setup(struct snd_sof_dev *sdev, struct snd_sof_pcm *spcm, in
 		if (!swidget)
 			continue;
 
-		/*
-		 * The scheduler widget for a pipeline is not part of the connected DAPM
-		 * widget list and it needs to be set up before the widgets in the pipeline
-		 * are set up. The use_count for the scheduler widget is incremented for every
-		 * widget in a given pipeline to ensure that it is freed only after the last
-		 * widget in the pipeline is freed.
-		 */
-		pipe_widget = swidget->pipe_widget;
-		if (!pipe_widget) {
-			dev_err(sdev->dev, "error: no pipeline widget found for %s\n",
-				swidget->widget->name);
-			ret = -EINVAL;
-			goto widget_free;
-		}
-
-		ret = sof_widget_setup(sdev, pipe_widget);
-		if (ret < 0)
-			goto widget_free;
-
```

Comment:
don't you need to remove pipe_widget line 442?

But pipe_widget is also used line 450 without ever having been initialized so something's very odd here.

also does this mean that the pipeline has to be first widget in the for_each_dapm_widget() loop? is this a valid assumption to make?

Reply:
yes, you are right about pipe_widget. I removed pipe_widget now. And no pipeline is not at all in the widget_list, so for_each_dapm_widget() doesnt contain it. Every widget has its pipe_widget field set and thats how we know which one to set